File: stack.java


public class Stack<T> {

    private T[] stackArray;
    private int top;
    private int maxSize;

   
    public Stack(int size) {
        stackArray = (T[]) new Object[size]; // 제네릭 배열 생성
        maxSize = size;
        top = -1;
    }

    
    public void push(T item) {
        if (top >= maxSize - 1) {
            System.out.println("Stack overflow");
        } else {
            stackArray[++top] = item;
        }
    }

    
    public T pop() {
        if (top < 0) {
            System.out.println("Stack underflow");
            return null;
        } else {
            return stackArray[top--];
        }
    }

    
    public T peek() {
        if (top < 0) {
            System.out.println("Stack is empty");
            return null;
        } else {
            return stackArray[top];
        }
    }

    
    public boolean isEmpty() {
        return top < 0;
    }

    // 스택의 크기 반환
    public int size() {
        return top + 1;
    }

    public static void main(String[] args) {
        
        Stack<Integer> intStack = new Stack<>(5);
        intStack.push(10);
        intStack.push(20);
        intStack.push(30);

        System.out.println("Peek (int): " + intStack.peek());
        System.out.println("Pop (int): " + intStack.pop()); 
        System.out.println("Size (int): " + intStack.size());  

        
        Stack<String> stringStack = new Stack<>(5);
        stringStack.push("Hello");
        stringStack.push("World");

        System.out.println("Peek (string): " + stringStack.peek()); 
        System.out.println("Pop (string): " + stringStack.pop());    
        System.out.println("Size (string): " + stringStack.size());
    }
}

File: stack.cs
using System;

class CustomStack<T>
{
    private T[] stackArray;
    private int top;
    private int maxSize;

    
    public CustomStack(int size)
    {
        stackArray = new T[size];
        maxSize = size;
        top = -1;
    }

    
    public void Push(T item)
    {
        if (top >= maxSize - 1)
        {
            Console.WriteLine("Stack overflow");
        }
        else
        {
            stackArray[++top] = item;
        }
    }

    
    public T Pop()
    {
        if (top < 0)
        {
            Console.WriteLine("Stack underflow");
            return default(T);
        }
        else
        {
            return stackArray[top--];
        }
    }

    // 스택의 맨 위 항목 확인
    public T Peek()
    {
        if (top < 0)
        {
            Console.WriteLine("Stack is empty");
            return default(T);
        }
        else
        {
            return stackArray[top];
        }
    }

    
    public bool IsEmpty()
    {
        return top < 0;
    }

   
    public int Size()
    {
        return top + 1;
    }
}

class Program
{
    static void Main()
    {
        
        CustomStack<int> intStack = new CustomStack<int>(5);
        intStack.Push(10);
        intStack.Push(20);
        intStack.Push(30);

        Console.WriteLine("Peek (int): " + intStack.Peek());  // 30
        Console.WriteLine("Pop (int): " + intStack.Pop());    // 30
        Console.WriteLine("Size (int): " + intStack.Size());  // 2

        
        CustomStack<string> stringStack = new CustomStack<string>(5);
        stringStack.Push("Hello");
        stringStack.Push("World");

        Console.WriteLine("Peek (string): " + stringStack.Peek());  // World
        Console.WriteLine("Pop (string): " + stringStack.Pop());    // World
        Console.WriteLine("Size (string): " + stringStack.Size());  // 1

        
        CustomStack<object> objectStack = new CustomStack<object>(5);
        objectStack.Push(1);
        objectStack.Push("Stack");
        objectStack.Push(3.14);

        Console.WriteLine("Peek (object): " + objectStack.Peek());  // 3.14
        Console.WriteLine("Pop (object): " + objectStack.Pop());    // 3.14
        Console.WriteLine("Size (object): " + objectStack.Size());  // 2
    }
}